x-pack/legacy/plugins/ml/public/jobs/new_job_new/recognize/components/job_settings_form.tsx

```diff
@@ -0,0 +1,320 @@
+import { TimeRangePicker } from '../../pages/components/time_range_step/time_range_picker';
+import { ModuleJobUI, SAVE_STATE } from '../page';
+import { getTimeFilterRange } from '../../../../components/full_time_range_selector';
+import { useKibanaContext } from '../../../../contexts/kibana';
+import {
+  composeValidators,
+  maxLengthValidator,
+  patternValidator,
+} from '../../../../../common/util/validators';
+import { JOB_ID_MAX_LENGTH } from '../../../../../common/constants/validation';
+import { isJobIdValid } from '../../../../../common/util/job_utils';
+import { TimeRange } from '../../pages/components/time_range_step/time_range';
+import { usePartialState } from '../../../../components/custom_hooks';
+
+export interface JobSettingsFormValues {
+  jobPrefix: string;
+  startDatafeedAfterSave: boolean;
+  useFullIndexData: boolean;
+  timeRange: TimeRange;
+  useDedicatedIndex: boolean;
+  jobGroups: string[];
+}
+
+interface JobSettingsFormProps {
+  saveState: SAVE_STATE;
+  onSubmit: (values: JobSettingsFormValues) => any;
+  onChange: (values: JobSettingsFormValues) => any;
+  jobGroups: string[];
+  existingGroupIds: string[];
+  jobs: ModuleJobUI[];
+}
+
+export const JobSettingsForm: FC<JobSettingsFormProps> = ({
+  onSubmit,
+  onChange,
+  saveState,
+  existingGroupIds,
+  jobs,
+}) => {
+  const { from, to } = getTimeFilterRange();
+  const { currentIndexPattern: indexPattern } = useKibanaContext();
+
+  const jobPrefixValidator = composeValidators(
+    patternValidator(/^([a-z0-9]+[a-z0-9\-_]*)?$/),
+    maxLengthValidator(JOB_ID_MAX_LENGTH - Math.max(...jobs.map(({ id }) => id.length)))
+  );
+  const groupValidator = composeValidators(
+    (value: string) => (isJobIdValid(value) ? null : { pattern: true }),
+    maxLengthValidator(JOB_ID_MAX_LENGTH)
+  );
+
+  const [formState, setFormState] = usePartialState({
+    jobPrefix: '',
+    startDatafeedAfterSave: true,
+    useFullIndexData: true,
+    timeRange: {
+      start: from,
+      end: to,
+    },
+    useDedicatedIndex: false,
+    jobGroups: [] as string[],
+  });
+  const [validationResult, setValidationResult] = useState<any>({});
+
+  const onJobPrefixChange = (value: string) => {
+    setFormState({
+      jobPrefix: value && value.toLowerCase(),
+    });
+  };
+
+  const handleValidation = () => {
+    const jobPrefixValidationResult = jobPrefixValidator(formState.jobPrefix);
+    const jobGroupsValidationResult = formState.jobGroups
+      .map(group => groupValidator(group))
+      .filter(result => result !== null);
+
+    setValidationResult({
+      jobPrefix: jobPrefixValidationResult,
+      jobGroups: jobGroupsValidationResult,
+      formValid: !jobPrefixValidationResult && jobGroupsValidationResult.length === 0,
+    });
+  };
+
+  useEffect(() => {
+    handleValidation();
+  }, [formState.jobPrefix, formState.jobGroups]);
+
+  useEffect(() => {
+    onChange(formState);
+  }, [formState]);
```

Comment:
Are you sure this works as expected? We need to be careful with objects and useEffect's shallow comparison (see https://github.com/facebook/react/issues/14476#issuecomment-471199055)